File: apps/client/src/shell/panels/git-actions.ts
import i18next from 'i18next';
import type { GitActionKind, GitActionPhase, GitStatus } from '@ruimte/contracts';

let counter = 0;

/* Names one run, so the progress of this push is not drawn on the toast of the one before it. */
export const nextActionId = (): string => {
    counter += 1;
    return `git-${Date.now()}-${counter}`;
};

export interface PushButton {
    label: string;
    kind: GitActionKind;
    disabled: boolean;
    /* What the tooltip says, which on a disabled button is why it is one. */
    reason: string;
}

/*
 * The primary button of the header. A branch git has never seen is published with an upstream in
 * the same push, which is a different word for the person and a different flag for git, so the
 * button says which of the two it is instead of failing halfway.
 */
export const pushButton = (status: GitStatus | null): PushButton => {
    const push = i18next.t('panels:git.push.label');
    if (status === null || !status.repo) {
        return { label: push, kind: 'push', disabled: true, reason: i18next.t('panels:git.push.noRepo') };
    }
    if (status.detached || status.branch === null) {
        return { label: push, kind: 'push', disabled: true, reason: i18next.t('panels:git.push.detached') };
    }
    if (status.upstream === null) {
        return {
            label: i18next.t('panels:git.push.publish'),
            kind: 'publish',
            disabled: false,
            reason: i18next.t('panels:git.push.publishReason', { branch: status.branch })
        };
    }
    if (status.ahead === 0) {
        return { label: push, kind: 'push', disabled: true, reason: i18next.t('panels:git.push.nothing') };
    }
    return { label: push, kind: 'push', disabled: false, reason: i18next.t('panels:git.push.reason', { count: status.ahead, upstream: status.upstream }) };
};

/* What the push menu offers for one repository: what its own button would say and how far ahead it is. */
export interface PushEntry {
    cwd: string;
    label: string;
    button: PushButton;
    ahead: number;
}

export const pushEntries = (checkouts: readonly { path: string; label: string; status: GitStatus | null }[]): PushEntry[] =>
    checkouts.map((checkout) => ({
        cwd: checkout.path,
        label: checkout.label,
        button: pushButton(checkout.status),
        ahead: checkout.status?.ahead ?? 0
    }));

/* The repositories a push would actually move, which is what "push all" runs and counts. */
export const pushable = (entries: readonly PushEntry[]): PushEntry[] => entries.filter((entry) => !entry.button.disabled);

/*
 * The primary button while the folder holds more than one repository. It pushes all of them: pushing
 * one of nine is the exception, and the flyout beside it is where that one is. Nothing to push
 * anywhere disables the button and the flyout with it, since every row in there would be dead too.
 */
export const pushAllButton = (entries: readonly PushEntry[]): PushButton => {
    const ready = pushable(entries);
    return {
        label: i18next.t('panels:git.push.allLabel'),
        kind: 'push',
        disabled: ready.length === 0,
        reason: ready.length === 0 ? i18next.t('panels:git.push.nothing') : i18next.t('panels:git.push.allReason', { count: ready.length })
    };
};

/* What the toast says while a run over several repositories is on this one. A kind this version has
   no words for still counts its way through rather than showing the name of a missing key. */
export const manyTitle = (kind: GitActionKind, repo: string, done: number, total: number): string =>
    i18next.t(`panels:git.many.${kind}`, { repo, done, total, defaultValue: i18next.t('panels:git.many.working', { repo, done, total }) });

/* What it says once that run is over. A repository that failed does not stop the others, so the end
   is the only place the whole outcome can be read. */
export const manySummary = (done: number, failed: readonly string[]): string =>
    failed.length === 0
        ? i18next.t('panels:git.many.done', { count: done })
        : i18next.t('panels:git.many.failed', { count: failed.length, repos: failed.join(', ') });

/* One checkout as the commit box weighs it: what is staged in it decides whether it takes the commit. */
export interface CommitCandidate {
    path: string;
    label: string;
    status: GitStatus | null;
}

/*
 * Where a commit would land. Every repository with something staged takes it, which is how a person
 * says with the files themselves what belongs in one commit. Nothing staged anywhere and exactly one
 * repository with changes is the commit that stages that repository first, the way it always was;
 * with more than one, staging is what has to say which of them is meant, so there is nothing to
 * commit yet.
 */
export const commitTargets = (checkouts: readonly CommitCandidate[]): { targets: CommitCandidate[]; stageAll: boolean } => {
    const staged = checkouts.filter((checkout) => checkout.status?.files.some((file) => file.state === 'staged') === true);
    if (staged.length > 0) {
        return { targets: staged, stageAll: false };
    }
    const changed = checkouts.filter((checkout) => (checkout.status?.files.length ?? 0) > 0);
    return changed.length === 1 ? { targets: changed, stageAll: true } : { targets: [], stageAll: false };
};

/* The title of the toast while an action runs: where it is, in the words a person would use. */
export const phaseLabel = (phase: GitActionPhase): string => i18next.t(`panels:git.phase.${phase}`);

/* What the toast says the moment an action starts, before git has written a line. A kind this
   version has no words for still gets a title rather than the name of a missing key. */
export const actionTitle = (kind: GitActionKind): string => i18next.t(`panels:git.action.${kind}`, { defaultValue: i18next.t('panels:git.action.working') });

/* The first line is the subject and the rest is the body, the way git itself reads a message. */
export const splitMessage = (message: string): { subject: string; body: string } => {
    const [subject = '', ...rest] = message.split('\n');
    return { subject: subject.trim(), body: rest.join('\n').trim() };
};

/* Git refuses to delete a branch it has not merged anywhere; only that refusal earns a second ask. */
export const isUnmergedRefusal = (message: string): boolean => message.includes('not fully merged');

File: apps/ios/App/Pages/Git/GitActions.swift
import RuimtePulsar
import RuimteTransport
import SwiftUI

/// What a person can be asked to confirm before it runs, and what the sheet for it says.
enum GitConfirmation: Identifiable {
    case discard(cwd: String, path: String)
    case forcePush(cwd: String)
    case deleteBranch(cwd: String, ref: String, force: Bool)
    case checkout(cwd: String, ref: String)

    var id: String {
        switch self {
        case .discard(let cwd, let path): "discard\u{0}\(cwd)\u{0}\(path)"
        case .forcePush(let cwd): "force-push\u{0}\(cwd)"
        case .deleteBranch(let cwd, let ref, let force): "delete\u{0}\(cwd)\u{0}\(ref)\u{0}\(force)"
        case .checkout(let cwd, let ref): "checkout\u{0}\(cwd)\u{0}\(ref)"
        }
    }

    var title: String {
        switch self {
        case .discard(_, let path): "Discard changes in \((path as NSString).lastPathComponent)?"
        case .forcePush: "Force push this branch?"
        case .deleteBranch(_, let ref, let force): force ? "Delete \(ref) anyway?" : "Delete \(ref)?"
        case .checkout(_, let ref): "Switch to \(ref)?"
        }
    }

    var detail: String {
        switch self {
        case .discard: "The file goes back to what it was, and git keeps a stash to undo it with."
        case .forcePush: "The remote branch is replaced by this one. Commits only it has are lost."
        case .deleteBranch(_, _, let force):
            force ? "This branch has not been merged anywhere. Its commits are only reachable by hash." : "The branch is removed here; the remote keeps its own."
        case .checkout: "Changes in the working tree are stashed first, so nothing is lost."
        }
    }

    var confirmLabel: String {
        switch self {
        case .discard: "Discard"
        case .forcePush: "Force push"
        case .deleteBranch: "Delete"
        case .checkout: "Switch"
        }
    }
}

extension GitRepositories {
    /// One action on one repository. The lines git writes while it runs land in `progress`, so a slow
    /// push says where it is instead of freezing a button.
    @discardableResult func act(
        client: any MachineRequesting, cwd: String, kind: String, extra: [String: JSONValue] = [:]
    ) async -> JSONValue? {
        guard !busy else { return nil }
        busy = true
        defer {
            busy = false
            progress = nil
        }
        switch await perform(client: client, cwd: cwd, kind: kind, extra: extra) {
        case .success(let result):
            problem = nil
            await refresh(client: client, cwd: cwd)
            return result
        case .failure(let error):
            problem = message(of: error)
            await refresh(client: client, cwd: cwd)
            return nil
        }
    }

    /// Every repository in order, never two at once: a folder of nine pushes over one link. A repository
    /// that fails does not stop the run, so the end is the only place the whole outcome can be read.
    func actAll(client: any MachineRequesting, kind: String) async {
        guard !busy, !checkouts.isEmpty else { return }
        busy = true
        defer {
            busy = false
            progress = nil
        }
        var failed: [String] = []
        for (index, checkout) in checkouts.enumerated() {
            step = "\(checkout.label) (\(index + 1)/\(checkouts.count))"
            if case .failure(let error) = await perform(client: client, cwd: checkout.path, kind: kind, extra: [:]) {
                failed.append("\(checkout.label): \(message(of: error))")
            }
        }
        step = nil
        problem = failed.isEmpty ? nil : failed.joined(separator: "\n")
        await refreshAll(client: client)
    }

    /// One message over every repository that has something staged: one commit each, the same words.
    func commit(
        client: any MachineRequesting, targets: [GitCheckout], subject: String, body: String, stageAll: Bool,
        push: Bool
    ) async -> Bool {
        guard !busy, !targets.isEmpty else { return false }
        busy = true
        defer {
            busy = false
            progress = nil
            step = nil
        }
        let kind = push ? "commit-push" : "commit"
        let extra: [String: JSONValue] = [
            "subject": .string(subject), "body": .string(body), "stageAll": .bool(stageAll),
        ]
        var failed: [String] = []
        for (index, target) in targets.enumerated() {
            step = targets.count > 1 ? "\(target.label) (\(index + 1)/\(targets.count))" : nil
            if case .failure(let error) = await perform(client: client, cwd: target.path, kind: kind, extra: extra) {
                failed.append("\(target.label): \(message(of: error))")
            }
        }
        problem = failed.isEmpty ? nil : failed.joined(separator: "\n")
        await refreshAll(client: client)
        return failed.isEmpty
    }

    /// Moves paths of one repository in or out of its index.
    func stage(client: any MachineRequesting, cwd: String, paths: [String], staged: Bool) async {
        guard !busy, !paths.isEmpty else { return }
        busy = true
        defer { busy = false }
        do {
            _ = try await client.request(
                "git.stage",
                payload: .object([
                    "cwd": .string(cwd), "paths": .array(paths.map { .string($0) }), "staged": .bool(staged),
                ]))
            problem = nil
        } catch {
            problem = message(of: error)
        }
        await refresh(client: client, cwd: cwd)
    }

    /// Throws away what a file holds, behind the stash git keeps to undo it with.
    func discard(client: any MachineRequesting, cwd: String, path: String) async -> String? {
        guard !busy else { return nil }
        busy = true
        defer { busy = false }
        var stash: String?
        do {
            let answer = try await client.request(
                "git.discard", payload: .object(["cwd": .string(cwd), "paths": .array([.string(path)])]))
            stash = answer["stash"]?.stringValue
            problem = nil
        } catch {
            problem = message(of: error)
        }
        await refresh(client: client, cwd: cwd)
        return stash
    }

    private func perform(
        client: any MachineRequesting, cwd: String, kind: String, extra: [String: JSONValue]
    ) async -> Result<JSONValue, any Error> {
        let actionId = UUID().uuidString
        let cancel = client.subscribe("git.progress") { [weak self] payload in
            guard payload.text("actionId") == actionId else { return }
            let line = payload.text("line")
            self?.progress = line.isEmpty ? payload.text("phase") : line
        }
        defer {
            cancel()
            progress = nil
        }
        var fields: [String: JSONValue] = [
            "cwd": .string(cwd), "actionId": .string(actionId), "kind": .string(kind),
        ]
        for (key, value) in extra { fields[key] = value }
        do {
            return .success(try await client.request("git.action", payload: .object(fields)))
        } catch {
            return .failure(error)
        }
    }

    private func message(of error: any Error) -> String {
        error.localizedDescription
    }
}

/// Git refuses to delete a branch it has not merged anywhere; only that refusal earns a second ask.
func gitIsUnmergedRefusal(_ message: String) -> Bool { message.contains("not fully merged") }
